interpret: dyn trait metadata check: equate traits in a proper way

RalfJung opened the pull request:
Hopefully fixes https://github.com/rust-lang/miri/issues/3541... unfortunately we don't have a testcase.

The first commit is just a refactor without functional change.

r? @oli-obk

RalfJung commented on compiler/rustc_const_eval/src/interpret/traits.rs:
```diff
@@ -57,4 +60,82 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+        if let (Some(expected_trait), Some(vtable_trait)) =
+            (expected_trait.principal(), vtable_trait)
+        {
+            // Slow path: spin up an inference context to check if these traits are sufficiently equal.
+            let infcx = self.tcx.infer_ctxt().build();
+            let ocx = ObligationCtxt::new(&infcx);
+            let cause = ObligationCause::dummy_with_span(self.cur_span());
+            // equate the two trait refs
+            if ocx.eq(&cause, self.param_env, expected_trait, vtable_trait).is_ok() {
+                if ocx.select_all_or_error().is_empty() {
+                    // All good.
+                    return Ok(());
+                }
+            }
```
@lcnr I hope this is the right way to do the trait equality check.

lcnr replied:
yes, that looks good to me :+1:

RalfJung commented:
@lcnr turns out that using the trait system here is not sufficient, the test still fails with this PR...

RalfJung commented:
Ah never mind, I missed one place where we are using `!=`.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9450569358/job/26029501701) [(plain)](https://github.com/rust-lang/rust/commit/10db95140ff8879bf7ef80e68fdf128fd93218a9/checks/26029501701/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     Running tests/ui.rs (obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-ff8e5f7317669711)
## Running ui tests in tests/pass for x86_64-unknown-linux-gnu
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-d71Eb2" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" OUT_DIR

FAILED TEST: tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-d71Eb2" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/ui/tests/pass/issues" "tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.stderr` to the actual output
--- tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.stderr
+++ <stderr output>
+error: Undefined Behavior: constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`
+  --> $DIR/issue-miri-3541-dyn-vtable-trait-normalization.rs:LL:CC
+   |
+LL |         unsafe { &mut *((emplacer_fn as *mut EmplacerFn<'a, T>) as *mut Self) }
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`
+   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
+   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
+   = note: BACKTRACE:
+   = note: BACKTRACE:
+   = note: inside `Emplacer::<'_, [std::boxed::Box<i32>]>::from_fn` at $DIR/issue-miri-3541-dyn-vtable-trait-normalization.rs:LL:CC
+note: inside `box_new_with::<[std::boxed::Box<i32>]>`
+  --> $DIR/issue-miri-3541-dyn-vtable-trait-normalization.rs:LL:CC
+   |
+LL |     unsafe { Emplacer::<T>::from_fn(emplacer_closure) };
+note: inside `main`
+  --> $DIR/issue-miri-3541-dyn-vtable-trait-normalization.rs:LL:CC
+   |
+   |
+LL |     box_new_with::<[Box<i32>]>();
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to 1 previous error
+error: aborting due to 1 previous error
+


error: there were 1 unmatched diagnostics
##[error]  --> tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:23:24
   |
23 |         unsafe { &mut *((emplacer_fn as *mut EmplacerFn<'a, T>) as *mut Self) }
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Error: Undefined Behavior: constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`

full stderr:
full stderr:
error: Undefined Behavior: constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`
##[error]  --> tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:23:24
   |
LL |         unsafe { &mut *((emplacer_fn as *mut EmplacerFn<'a, T>) as *mut Self) }
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: BACKTRACE:
   = note: BACKTRACE:
   = note: inside `Emplacer::<'_, [std::boxed::Box<i32>]>::from_fn` at tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:23:24: 23:78
note: inside `box_new_with::<[std::boxed::Box<i32>]>`
  --> tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:35:14
   |
LL |     unsafe { Emplacer::<T>::from_fn(emplacer_closure) };
note: inside `main`
  --> tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:39:5
   |
   |
LL |     box_new_with::<[Box<i32>]>();

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
---
Error: 
   0: ui tests in tests/pass for x86_64-unknown-linux-gnu failed
   1: tests failed

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-ff8e5f7317669711 --quiet` (exit status: 1)
Build completed unsuccessfully in 0:03:13
  local time: Mon Jun 10 15:11:29 UTC 2024
```

</details>

RalfJung committed:
interpret: refactor dyn trait handling

We can check that the vtable is for the right trait very early, and then just pass the type around.

RalfJung committed:
interpret: dyn trait metadata check: equate traits in a proper way

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9451199788/job/26031611244) [(plain)](https://github.com/rust-lang/rust/commit/18c958d6abb182894d387bd91886a91a0f72b228/checks/26031611244/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     Running tests/ui.rs (obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-ff8e5f7317669711)
## Running ui tests in tests/pass for x86_64-unknown-linux-gnu
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-MMXQdT" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" OUT_DIR

FAILED TEST: tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-MMXQdT" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/ui/tests/pass/issues" "tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs" "--edition" "2021"
error: pass test got exit status: 1, but expected 0

error: actual output differed from expected
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.stderr` to the actual output
--- tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.stderr
+++ <stderr output>
+*not* equal: Binder { value: std::ops::FnMut<(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata,)>, bound_vars: [Region(BrNamed(DefId(0:9 ~ issue_miri_3541_dyn_vtable_trait_normalization[e867]::EmplacerFn::'b), 'b))] } Binder { value: std::ops::FnMut<(usize,)>, bound_vars: [] }
+error: Undefined Behavior: constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`
+  --> $DIR/issue-miri-3541-dyn-vtable-trait-normalization.rs:LL:CC
+   |
+LL |         unsafe { &mut *((emplacer_fn as *mut EmplacerFn<'a, T>) as *mut Self) }
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`
+   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
+   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
+   = note: BACKTRACE:
+   = note: BACKTRACE:
+   = note: inside `Emplacer::<'_, [std::boxed::Box<i32>]>::from_fn` at $DIR/issue-miri-3541-dyn-vtable-trait-normalization.rs:LL:CC
+note: inside `box_new_with::<[std::boxed::Box<i32>]>`
+  --> $DIR/issue-miri-3541-dyn-vtable-trait-normalization.rs:LL:CC
+   |
+LL |     unsafe { Emplacer::<T>::from_fn(emplacer_closure) };
+note: inside `main`
+  --> $DIR/issue-miri-3541-dyn-vtable-trait-normalization.rs:LL:CC
+   |
+   |
+LL |     box_new_with::<[Box<i32>]>();
+
+note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+
+error: aborting due to 1 previous error
+error: aborting due to 1 previous error
+


error: there were 1 unmatched diagnostics
##[error]  --> tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:23:24
   |
23 |         unsafe { &mut *((emplacer_fn as *mut EmplacerFn<'a, T>) as *mut Self) }
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Error: Undefined Behavior: constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`

full stderr:
full stderr:
*not* equal: Binder { value: std::ops::FnMut<(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata,)>, bound_vars: [Region(BrNamed(DefId(0:9 ~ issue_miri_3541_dyn_vtable_trait_normalization[e867]::EmplacerFn::'b), 'b))] } Binder { value: std::ops::FnMut<(usize,)>, bound_vars: [] }
error: Undefined Behavior: constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`
##[error]  --> tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:23:24
   |
LL |         unsafe { &mut *((emplacer_fn as *mut EmplacerFn<'a, T>) as *mut Self) }
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: wrong trait in wide pointer vtable: expected `std::ops::FnMut(<[std::boxed::Box<i32>] as std::ptr::Pointee>::Metadata)`, but encountered `std::ops::FnMut<(usize,)>`
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: BACKTRACE:
   = note: BACKTRACE:
   = note: inside `Emplacer::<'_, [std::boxed::Box<i32>]>::from_fn` at tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:23:24: 23:78
note: inside `box_new_with::<[std::boxed::Box<i32>]>`
  --> tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:35:14
   |
LL |     unsafe { Emplacer::<T>::from_fn(emplacer_closure) };
note: inside `main`
  --> tests/pass/issues/issue-miri-3541-dyn-vtable-trait-normalization.rs:39:5
   |
   |
LL |     box_new_with::<[Box<i32>]>();

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
---
Error: 
   0: ui tests in tests/pass for x86_64-unknown-linux-gnu failed
   1: tests failed

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-ff8e5f7317669711 --quiet` (exit status: 1)
Build completed unsuccessfully in 0:03:21
  local time: Mon Jun 10 15:51:29 UTC 2024
```

</details>

lcnr commented on compiler/rustc_const_eval/src/interpret/traits.rs:
```diff
@@ -33,28 +36,88 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+            let infcx = self.tcx.infer_ctxt().build();
+            let ocx = ObligationCtxt::new(&infcx);
+            let cause = ObligationCause::dummy_with_span(self.cur_span());
+            // equate the two trait refs
```
that's because I forgot that the old solver exists/didn't look carefully enough. We still need to first normalize here

```suggestion
            // equate the two trait refs after normalization
            let expected_trait = ocx.normalize(&cause, self.param_env, expected_trait);
            let vtable_trait = ocx.normalize(&cause, self.param_env, vtable_trait);
```

RalfJung commented:
So just normalizing and then using == does not work?

RalfJung commented:
Makes sense. This seems to work, thanks!

@rustbot ready

RalfJung committed:
check for correct trait in size_and_align_of

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9460271632/job/26058782451) [(plain)](https://github.com/rust-lang/rust/commit/08ee7031c5c0889d95a470349c64fb07828dbf1b/checks/26058782451/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
## Running ui tests in tests/fail for x86_64-unknown-linux-gnu
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-4lcBQj" "RUST_BACKTRACE"="1" /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" OUT_DIR

FAILED TEST: tests/fail/dyn-call-trait-mismatch.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-4lcBQj" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/ui/tests/fail" "tests/fail/dyn-call-trait-mismatch.rs" "-Zmiri-disable-validation" "--edition" "2021"
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/fail/dyn-call-trait-mismatch.stderr` to the actual output
--- tests/fail/dyn-call-trait-mismatch.stderr
+++ <stderr output>
+++ <stderr output>
 error: Undefined Behavior: using vtable for trait `T1` but trait `T2` was expected
   --> $DIR/dyn-call-trait-mismatch.rs:LL:CC
    |
-LL |     r2.method2();
+LL |     let r2: Box<dyn T2> = unsafe { std::mem::transmute(r) };
-   |     ^^^^^^^^^^^^ using vtable for trait `T1` but trait `T2` was expected
+   |                                    ^^^^^^^^^^^^^^^^^^^^^^ using vtable for trait `T1` but trait `T2` was expected
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
... 6 lines skipped ...
 error: aborting due to 1 previous error
 
 


error: `using vtable for trait `T1` but trait `T2` was expected` not found in diagnostics on line 19
   |
   |
19 |     r2.method2(); //~ERROR: using vtable for trait `T1` but trait `T2` was expected
   |

error: there were 1 unmatched diagnostics
##[error]  --> tests/fail/dyn-call-trait-mismatch.rs:18:36
##[error]  --> tests/fail/dyn-call-trait-mismatch.rs:18:36
   |
18 |     let r2: Box<dyn T2> = unsafe { std::mem::transmute(r) };
   |                                    ^^^^^^^^^^^^^^^^^^^^^^ Error: Undefined Behavior: using vtable for trait `T1` but trait `T2` was expected

full stderr:
error: Undefined Behavior: using vtable for trait `T1` but trait `T2` was expected
##[error]  --> tests/fail/dyn-call-trait-mismatch.rs:18:36
##[error]  --> tests/fail/dyn-call-trait-mismatch.rs:18:36
   |
LL |     let r2: Box<dyn T2> = unsafe { std::mem::transmute(r) };
   |                                    ^^^^^^^^^^^^^^^^^^^^^^ using vtable for trait `T1` but trait `T2` was expected
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: BACKTRACE:
   = note: inside `main` at tests/fail/dyn-call-trait-mismatch.rs:18:36: 18:58
---
Error: 
   0: ui tests in tests/fail for x86_64-unknown-linux-gnu failed
   1: tests failed

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-ff8e5f7317669711 --quiet` (exit status: 1)
Build completed unsuccessfully in 0:03:57
  local time: Tue Jun 11 06:12:28 UTC 2024
```

</details>

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 3757136d8e0d8ddca294453e5a5ce70cfa3417e9 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 3757136d8e0d8ddca294453e5a5ce70cfa3417e9 -->
<!-- homu: {"type":"Approved","sha":"3757136d8e0d8ddca294453e5a5ce70cfa3417e9","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
